solr/solrj/src/java/org/apache/solr/common/cloud/SolrZkClient.java

```diff
@@ -347,10 +350,20 @@ public List<String> getChildren(final String path, final Watcher watcher,Stat st
    */
   public byte[] getData(final String path, final Watcher watcher, final Stat stat, boolean retryOnConnLoss)
```

Comment:
I think we are only planning to compress `state.json` so , other files should not be paying a price for this check. This [method](https://github.com/fullstorydev/lucene-solr/blob/0cb904f1df2baa8ffa8c1b98f79304872940870c/solr/solrj/src/java/org/apache/solr/common/cloud/ClusterState.java#L236) is probably a better place to implement this

Reply:
> a very generic question.
> 
> As we have the ability to read both formats, should we just have a configuration to use compression if the size of `state.json > some configured size ` ?
> 
> Are the small collections going to pay a price unnecessarily ?

I've been going back and forth on this. Only compressing sometimes could add some confusion, but it is probably worth starting with only compression for state.json > 500k and we can iterate from there. I'll add that in.

Reply:
I originally was looking at implementing this in ClusterState.load() however this means that it would also have to be implemented in a variety of other places to make Zookeeper CLI and Zookeeper Admin API calls also work. The check to see if data is compressed is just checking 4 bytes of the data so I think it is lightweight enough that it won't add significantly to calls for non-compressed data.

I ran some tests to validate this and the compression check adds 0.01-0.03 milliseconds on average so pretty negligible time difference and makes the code easier.

Reply:
One problem I see with the current solution is we make 2 copies of the data. one compressed and one uncompressed. So, we end up using more memory in the nodes. The JSON parser can directly use a ZIpInputStream and parse if the original bytes are made available

Reply:
This is true, but iiuc I definitely wouldn't characterize this as a "problem with the current solution", and I'd scope the adjustment you're proposing as a separate follow-on change. Currently the JSON parser expects/gets the full uncompressed state.json as an in-memory `byte[]`; iiuc the only overhead this PR adds over the status quo is that jvm needs to hold the _compressed_ state.json in memory in a transient `byte[]`, which is released immediately once decompression is complete.

The read from zk is not streaming (in keeping with zk's admonition to keep data small anyway), so without this PR, parsing JSON from a stream as opposed to an in-memory `byte[]` would have been pointless. So unless I'm missing something, this PR is in fact a prerequisite for streaming parsing of state.json to be beneficial.

Reply:
We can directly open a ZipInputstream from a ByteArrayInputStream . That would've actually been a more memory efficient solution.

Anyway, this is not a deal breaker.

Reply:
In principle, yes. Unfortunately at the moment it seems that the assumption that you're dealing directly with utf8 `byte[]` is pretty well baked into all the code that calls `SolrZkClient.getData()`. Placing this change as close to the reading of the raw bytes from zk as possible feels to me like the cleanest way to go, at least initially. I'm sure it'd be possible, but not sure how straightforward to be 100% sure tracking down all the calling code that needs to be "compression-aware" and adapting it to handle decompression.

I hesitate to mention, b/c I'm not thrilled with the idea of increasing the API footprint, but: one possible way forward would be to leave "compression-awareness" in the current `byte[] SolrZkClient.getData()`, and implement an additional `InputStream ZolrZkClient.getData()` that returns a utf8 inputstream instead of utf8 bytes. The semantics would be pretty clear, and that way all callers that rely on the current `getData()` always providing normal utf8 bytes would still work as previously, but callers that want to take advantage of this possible optimization in the case of compressed input could "opt in" and call the version of `getData()` that returns a utf8 InputStream -- all compression logic still handled in one place.

In any event I do think this PR in its current state increases heap memory consumption only minimally and transiently. Compared to other memory consumption even in the immediate we're considering here, a 20% increase in raw bytes (assuming a ~5x compression ratio) is unlikely to be an issue I think?

Reply:
>its current state increases heap memory consumption only minimally and transiently. Compared to other memory consumption even in the immediate we're considering ...

But, you are assuming that the price , albeit not significant, is unavoidable. I was suggesting that it was avoidable. 

Your suggestion of another 'method getDatsStream()' that returns a stream is a good solution that eliminates the double copy .

Reply:
I like the idea of an additional method, but I'd like to take that to a separate PR if that works for you. If this one looks good otherwise, we can merge it and follow up later with additional improvements.

Reply:
sure. A separate PR would be OK. This can go in as is